yarn-project/simulator/src/avm/journal/journal.ts

```diff
@@ -122,10 +122,14 @@ export class AvmPersistableStateManager {
    * @returns true if the note hash exists at the given leaf index, false otherwise
    */
   public async checkNoteHashExists(storageAddress: Fr, noteHash: Fr, leafIndex: Fr): Promise<boolean> {
-    const gotLeafIndex = await this.hostStorage.commitmentsDb.getCommitmentIndex(noteHash);
-    const exists = gotLeafIndex === leafIndex.toBigInt();
-    this.log.debug(`noteHashes(${storageAddress})@${noteHash} ?? leafIndex: ${leafIndex}, exists: ${exists}.`);
-    this.trace.traceNoteHashCheck(storageAddress, noteHash, leafIndex, exists);
+    const gotLeafValue = await this.hostStorage.commitmentsDb.getCommitmentValue(leafIndex.toBigInt());
+    const exists = gotLeafValue?.equals(noteHash) ?? false;
+    this.log.debug(
+      `noteHashes(${storageAddress})@${noteHash} ?? leafIndex: ${leafIndex} | gotLeafValue: ${gotLeafValue}, exists: ${exists}.`,
+    );
+    // TODO(8287): We still return exists here, but we need to transmit both the requested noteHash and the gotLeafValue
+    // such that the VM can constrain the equality and decide on exists based on that.
+    this.trace.traceNoteHashCheck(storageAddress, gotLeafValue ?? noteHash, leafIndex, exists);
```

Comment:
@LeilaWang What should be the behaviour if `leafIndex` (and therefore `gotLeafValue`) is undefined?

Right now we return `noteHash` since it required the least amount of changes to tests 😅 , equally it could be Fr(0)?